src/useObserver.ts

```diff
@@ -24,42 +28,200 @@ export function useObserver<T>(
+                // got dropped, and we don't want to make state changes then.
+                // (It triggers warnings in StrictMode, for a start.)
+                if (trackingData.mounted) {
+                    // We have reached useEffect(), so we're mounted, and can trigger an update
+                    forceUpdate()
+                } else {
+                    // We haven't yet reached useEffect(), so we'll need to trigger a re-render
+                    // when (and if) useEffect() arrives.  The easiest way to do that is just to
+                    // drop our current reaction and allow useEffect() to recreate it.
+                    trackingData.reaction.dispose()
+                    reactionTrackingRef.current = null
+                }
+            })
+        }
+
+        reactionTrackingRef.current = trackingData
+        scheduleCleanupOfReactionIfLeaked(reactionTrackingRef)
     }
 
+    const reaction = reactionTrackingRef.current!.reaction
     useDebugValue(reaction, printDebugValue)
 
     useEffect(() => {
-        committed.current = true
-        return () => reaction.current!.dispose()
+        // Called on first mount only
+        recordReactionAsCommitted(reactionTrackingRef)
+
+        if (reactionTrackingRef.current) {
+            // Great. We've already got our reaction from our render;
+            // all we need to do is to record that it's now mounted,
+            // to allow future observable changes to trigger re-renders
+            reactionTrackingRef.current.mounted = true
+        } else {
+            // The reaction we set up in our render has been disposed.
+            // This is either due to bad timings of renderings, e.g. our
+            // component was paused for a _very_ long time, and our
+            // reaction got cleaned up, or we got a observable change
+            // between render and useEffect
+
+            // Re-create the reaction
+            reactionTrackingRef.current = {
+                reaction: new Reaction(observerComponentNameFor(baseComponentName), () => {
+                    // We've definitely already been mounted at this point
+                    forceUpdate()
+                }),
+                cleanAt: Infinity
+            }
+            forceUpdate()
+        }
+
+        return () => reactionTrackingRef.current!.reaction.dispose()
     }, [])
 
     // render the original component, but have the
     // reaction track the observables, so that rendering
     // can be invalidated (see above) once a dependency changes
     let rendering!: T
     let exception
-    reaction.current.track(() => {
+    reaction.track(() => {
         try {
             rendering = fn()
         } catch (e) {
             exception = e
         }
     })
     if (exception) {
-        reaction.current.dispose()
         throw exception // re-throw any exceptions catched during rendering
     }
     return rendering
 }
+
```

Comment:
Nit: could be nice to move this logic and state to a separate file?

Reply:
Yeah, that would be nicer. I'll get on it.

Reply:
Done.